jobrunner/executors/local.py

```diff
@@ -313,11 +346,23 @@ def finalize_job(job_definition):
         container_name(job_definition), none_if_not_exists=True
     )
     if not container_metadata:
-        raise LocalDockerError("Job container has vanished")
+        if job_definition.cancelled:
+            # no logs to retain if the container didn't start yet
+            return
+        else:
+            raise LocalDockerError(
+                f"Job container {container_name(job_definition)} has vanished"
+            )
     redact_environment_variables(container_metadata)
 
-    outputs, unmatched_patterns = find_matching_outputs(job_definition)
-    unmatched_outputs = get_unmatched_outputs(job_definition, outputs)
+    if job_definition.cancelled:
+        # assume no outputs because our job didn't finish
+        outputs = {}
```

Comment:
nb. something to double-check - if someone accidentally re-runs a job & then cancels it, this is going to blank out their previous results? but that's correct as we'll also be overwriting the log file?

Reply:
If there are no outputs from the job, nothing will be copied from the volume to the workspace, so any previous outputs should still be there?

Only time they'd be overwritten is if they job exited with a 0 exit code, I think.

I think this goes back to the discussion with Dave a while back, where I thought we'd decided *not* to copy to the cancelled job's log file over the current `workspace/metadata/$action.log`, as the assumption was that they most likely don't want that behaviour.

I think we talked about copying it to another filename in that dir, perhaps? But I don't recall what we decided?

Reply:
this question still needs addressing

Reply:
ok, I think this was your (@bloodearnest) suggestion on Wednesday, but have proposed in this PR that we only do the store to log_dir & decide on filenames for how to copy to workspaces in a separate ticket - wdyt?

Reply:
wfm